docs/rfc/012-record-transaction-model.md

```diff
@@ -181,104 +197,171 @@ its own deserialization of the payload bytes into higher-level Leo values;
-// using an annotation for concreteness, but the point stands for the other options discussed
+@entrypoint
+function ...(input: TransactionInput, ...) -> TransactionOutput { ... }
 ```
-In other words, an entry point function can be now seen as a mathematical function
+This way, the calculation of transaction outputs from transaction inputs is made functional and explicit.
+
+As special cases (both of which may apply to the same entry point):
+1. We could allow the `TransactionInput` input to be absent,
+   when an entry point does not need access the transaction input data,
+   e.g. when producing new records without consuming old records.
+2. We could allow the function output to be `()` instead of `TransactionOutput`,
+   when an entry point does not need to produce transaction outputs,
+   e.g. when consuming old records without producing new records.
+
+Compared to the current Leo program execution model (described earlier in the background section),
+`input` is made an explicit input here, instead of being like a built-in global variable.
+Furthermore, the output type is restricted to be `TransactionOutput` (or `()`),
+thus eliminating the implicit serialization and the asymmetry with the treatment of transaction inputs.
+There is still no restriction on the non-`TransactionInput` inputs of an entry point function;
+as noted earlier, they are existentially quantified in the zero-knowledge assertion.
+
+Thus, a Leo program entry point can be now described as a mathematical function
 ```
 entrypoint : Record x Record x Inputs -> Record x Record
 ```
-where one or both output records are dummy if the function creates less than two new records.
-
-The above constrains each entry point to always return the same number of records.
-Different entry point functions may return different numbers of records.
-If we want the same entry point function
-to return different numbers of records in different situations,
-then it could make sense to have a more general circuit type for the output of a transaction,
-which may contain 0, 1, or 2 records, and possibly other information as needed,
-and require entry point functions to uniformly return values of that type:
-```
-@entrypoint function ...(...) -> TransactionOutput // contains 0, 1, or 2 records
-```
+where `Output` is no longer present.
+(If `TransactionInput` includes additional data, besides the two old records, that may affect the transaction output,
+then we would need to add that to this mathematical model;
+however, the model above is sufficiently accurate for the current discussion.)
+
+We may require the `TransactionInput` input of an entry point function, if present,
+to be the first input of the function, for clarity and readability.
+A question is whether we should extend that requirement to non-entry-point functions
+that may be passed `TransactionInput` values.
+We note that none of these restrictions are necessary, though.
```

Comment:
Though it makes analysis simpler, more efficient, and arguably the resulting code is more readable for devs also.